lab_1/main.py

```diff
@@ -3,184 +3,144 @@
-            top_words.append (key)
-        return top_words[:top_n-1]
-'''
-'''         if top_words == []:
-                top_words.append (key)
-            else:
-                for word in top_words:
-                    if value > freq_dict[word]:
-                        top_words = [key] + top_words
-                    else:
-                        top_words.append (key)
-    return top_words[:top_n-1]'''
-"""
+
+def get_top_n_words(freq_dict: dict, top_n: int) -> list:
+    """
     Returns the most common words
     :param freq_dict: a dictionary with frequencies
     :param top_n: a number of the most common words to return
     :return: a list of the most common words
     e.g. tokens = ['weather', 'sunny', 'man', 'happy', 'and', 'dog', 'happy']
     top_n = 1
     --> ['happy']
-"""
+    """
+    if type(freq_dict) != dict or type(top_n) != int:
+        return []
+    top_words = []
+    sorted_dict = {k: freq_dict[k] for k in sorted(freq_dict, key=freq_dict.get, reverse=True)}
+    for key in sorted_dict:
+        top_words.append(key)
+    return top_words[:top_n]
 
 
-def get_concordance(tokens, word, left_context_size, right_context_size):
+def get_concordance(tokens: list, word: str, left_context_size: int, right_context_size: int) -> list:
+    """
+    Gets a concordance of a word
+    A concordance is a listing of each occurrence of a word in a text,
+    presented with the words surrounding it
+    :param tokens: a list of tokens
+    :param word: a word-base for a concordance
+    :param left_context_size: the number of words in the left context
+    :param right_context_size: the number of words in the right context
+    :return: a concordance
+    e.g. tokens = ['the', 'weather', 'is', 'sunny', 'the', 'man', 'is', 'happy',
+                    'the', 'dog', 'is', 'happy', 'but', 'the', 'cat', 'is', 'sad']
+    word = 'happy'
+    left_context_size = 2
+    right_context_size = 3
+    --> [['man', 'is', 'happy', 'the', 'dog', 'is'], ['dog', 'is', 'happy', 'but', 'the', 'cat']]
+    """
     concordance = []
-    if type(tokens) != list or type(word) != str: # or word not in tokens
+    if type(tokens) != list or type(word) != str:
         return []
     elif type(left_context_size) != int or type(right_context_size) != int:
         return []
-    elif type(left_context_size) == int and type(right_context_size) == int and left_context_size < 1 and right_context_size < 1:
+    elif left_context_size < 1 and right_context_size < 1:
         return []
-    elif right_context_size < 1 and left_context_size >= 1:
-        for number, token in enumerate (tokens):
-            context = []
-            if token == word:
+    for number, token in enumerate(tokens):
+        context = []
+        if token == word:
+            if right_context_size < 1 and left_context_size >= 1:
                 left_size = left_context_size
                 while left_size != 0:
```

Comment:
can you explain, please, why do you use while here (and later)?

Reply:
I use it because I need to add words before the chosen word. If I use -for- to do it - it adds them in the wrong order, it makes the element before the chosen word first even when the left limit is more than one(for example, words go like that 3 2 1 -word-, but I get 1 2 3 -word-).  So I made a while-cycle which adds elements from the left first and then the chosen word (from big to small). It just looked easier for me to solve it like that)